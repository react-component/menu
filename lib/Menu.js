'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _createReactClass = require('create-react-class');

var _createReactClass2 = _interopRequireDefault(_createReactClass);

var _MenuMixin = require('./MenuMixin');

var _MenuMixin2 = _interopRequireDefault(_MenuMixin);

var _util = require('./util');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

// import React from 'react';
var Menu = (0, _createReactClass2["default"])({
  displayName: 'Menu',

  propTypes: {
    openSubMenuOnMouseEnter: _propTypes2["default"].bool,
    closeSubMenuOnMouseLeave: _propTypes2["default"].bool,
    selectedKeys: _propTypes2["default"].arrayOf(_propTypes2["default"].string),
    defaultSelectedKeys: _propTypes2["default"].arrayOf(_propTypes2["default"].string),
    defaultOpenKeys: _propTypes2["default"].arrayOf(_propTypes2["default"].string),
    openKeys: _propTypes2["default"].arrayOf(_propTypes2["default"].string),
    mode: _propTypes2["default"].string,
    onClick: _propTypes2["default"].func,
    onSelect: _propTypes2["default"].func,
    onDeselect: _propTypes2["default"].func,
    onDestroy: _propTypes2["default"].func,
    openTransitionName: _propTypes2["default"].string,
    openAnimation: _propTypes2["default"].oneOfType([_propTypes2["default"].string, _propTypes2["default"].object]),
    level: _propTypes2["default"].number,
    eventKey: _propTypes2["default"].string,
    selectable: _propTypes2["default"].bool,
    children: _propTypes2["default"].any
  },

  mixins: [_MenuMixin2["default"]],

  getDefaultProps: function getDefaultProps() {
    return {
      openSubMenuOnMouseEnter: true,
      closeSubMenuOnMouseLeave: true,
      selectable: true,
      onClick: _util.noop,
      onSelect: _util.noop,
      onOpenChange: _util.noop,
      onDeselect: _util.noop,
      defaultSelectedKeys: [],
      defaultOpenKeys: []
    };
  },
  getInitialState: function getInitialState() {
    var props = this.props;
    var selectedKeys = props.defaultSelectedKeys;
    var openKeys = props.defaultOpenKeys;
    if ('selectedKeys' in props) {
      selectedKeys = props.selectedKeys || [];
    }
    if ('openKeys' in props) {
      openKeys = props.openKeys || [];
    }
    return {
      selectedKeys: selectedKeys,
      openKeys: openKeys
    };
  },
  componentWillReceiveProps: function componentWillReceiveProps(nextProps) {
    var props = {};
    if ('selectedKeys' in nextProps) {
      props.selectedKeys = nextProps.selectedKeys || [];
    }
    if ('openKeys' in nextProps) {
      props.openKeys = nextProps.openKeys || [];
    }
    this.setState(props);
  },
  onDestroy: function onDestroy(key) {
    var state = this.state;
    var props = this.props;
    var selectedKeys = state.selectedKeys;
    var openKeys = state.openKeys;
    var index = selectedKeys.indexOf(key);
    if (!('selectedKeys' in props) && index !== -1) {
      selectedKeys.splice(index, 1);
    }
    index = openKeys.indexOf(key);
    if (!('openKeys' in props) && index !== -1) {
      openKeys.splice(index, 1);
    }
  },
  onItemHover: function onItemHover(e) {
    var item = e.item;
    var _props = this.props,
        mode = _props.mode,
        closeSubMenuOnMouseLeave = _props.closeSubMenuOnMouseLeave;
    var _e$openChanges = e.openChanges,
        openChanges = _e$openChanges === undefined ? [] : _e$openChanges;
    // special for top sub menu

    if (mode !== 'inline' && !closeSubMenuOnMouseLeave && item.isSubMenu) {
      var activeKey = this.state.activeKey;
      var activeItem = this.getFlatInstanceArray().filter(function (c) {
        return c && c.props.eventKey === activeKey;
      })[0];
      if (activeItem && activeItem.props.open) {
        openChanges = openChanges.concat({
          key: item.props.eventKey,
          item: item,
          originalEvent: e,
          open: true
        });
      }
    }
    openChanges = openChanges.concat(this.getOpenChangesOnItemHover(e));
    if (openChanges.length) {
      this.onOpenChange(openChanges);
    }
  },
  onSelect: function onSelect(selectInfo) {
    var props = this.props;
    if (props.selectable) {
      // root menu
      var selectedKeys = this.state.selectedKeys;
      var selectedKey = selectInfo.key;
      if (props.multiple) {
        selectedKeys = selectedKeys.concat([selectedKey]);
      } else {
        selectedKeys = [selectedKey];
      }
      if (!('selectedKeys' in props)) {
        this.setState({
          selectedKeys: selectedKeys
        });
      }
      props.onSelect((0, _extends3["default"])({}, selectInfo, {
        selectedKeys: selectedKeys
      }));
    }
  },
  onClick: function onClick(e) {
    this.props.onClick(e);
  },
  onOpenChange: function onOpenChange(e_) {
    var props = this.props;
    var openKeys = this.state.openKeys.concat();
    var changed = false;
    var processSingle = function processSingle(e) {
      var oneChanged = false;
      if (e.open) {
        oneChanged = openKeys.indexOf(e.key) === -1;
        if (oneChanged) {
          openKeys.push(e.key);
        }
      } else {
        var index = openKeys.indexOf(e.key);
        oneChanged = index !== -1;
        if (oneChanged) {
          openKeys.splice(index, 1);
        }
      }
      changed = changed || oneChanged;
    };
    if (Array.isArray(e_)) {
      // batch change call
      e_.forEach(processSingle);
    } else {
      processSingle(e_);
    }
    if (changed) {
      if (!('openKeys' in this.props)) {
        this.setState({ openKeys: openKeys });
      }
      props.onOpenChange(openKeys);
    }
  },
  onDeselect: function onDeselect(selectInfo) {
    var props = this.props;
    if (props.selectable) {
      var selectedKeys = this.state.selectedKeys.concat();
      var selectedKey = selectInfo.key;
      var index = selectedKeys.indexOf(selectedKey);
      if (index !== -1) {
        selectedKeys.splice(index, 1);
      }
      if (!('selectedKeys' in props)) {
        this.setState({
          selectedKeys: selectedKeys
        });
      }
      props.onDeselect((0, _extends3["default"])({}, selectInfo, {
        selectedKeys: selectedKeys
      }));
    }
  },
  getOpenTransitionName: function getOpenTransitionName() {
    var props = this.props;
    var transitionName = props.openTransitionName;
    var animationName = props.openAnimation;
    if (!transitionName && typeof animationName === 'string') {
      transitionName = props.prefixCls + '-open-' + animationName;
    }
    return transitionName;
  },
  isInlineMode: function isInlineMode() {
    return this.props.mode === 'inline';
  },
  lastOpenSubMenu: function lastOpenSubMenu() {
    var lastOpen = [];
    var openKeys = this.state.openKeys;

    if (openKeys.length) {
      lastOpen = this.getFlatInstanceArray().filter(function (c) {
        return c && openKeys.indexOf(c.props.eventKey) !== -1;
      });
    }
    return lastOpen[0];
  },
  renderMenuItem: function renderMenuItem(c, i, subIndex) {
    if (!c) {
      return null;
    }
    var state = this.state;
    var extraProps = {
      openKeys: state.openKeys,
      selectedKeys: state.selectedKeys,
      openSubMenuOnMouseEnter: this.props.openSubMenuOnMouseEnter
    };
    return this.renderCommonMenuItem(c, i, subIndex, extraProps);
  },
  render: function render() {
    var props = (0, _extends3["default"])({}, this.props);
    props.className += ' ' + props.prefixCls + '-root';
    return this.renderRoot(props);
  }
});

exports["default"] = Menu;
module.exports = exports['default'];