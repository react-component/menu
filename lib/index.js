'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Divider = exports.ItemGroup = exports.MenuItemGroup = exports.MenuItem = exports.Item = exports.SubMenu = undefined;

var _Menu = require('./Menu');

var _Menu2 = _interopRequireDefault(_Menu);

var _SubMenu = require('./SubMenu');

var _SubMenu2 = _interopRequireDefault(_SubMenu);

var _MenuItem = require('./MenuItem');

var _MenuItem2 = _interopRequireDefault(_MenuItem);

var _MenuItemGroup = require('./MenuItemGroup');

var _MenuItemGroup2 = _interopRequireDefault(_MenuItemGroup);

var _Divider = require('./Divider');

var _Divider2 = _interopRequireDefault(_Divider);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

exports.SubMenu = _SubMenu2["default"];
exports.Item = _MenuItem2["default"];
exports.MenuItem = _MenuItem2["default"];
exports.MenuItemGroup = _MenuItemGroup2["default"];
exports.ItemGroup = _MenuItemGroup2["default"];
exports.Divider = _Divider2["default"];
exports["default"] = _Menu2["default"];