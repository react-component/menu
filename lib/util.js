'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.noop = noop;
exports.getKeyFromChildrenIndex = getKeyFromChildrenIndex;
exports.loopMenuItem = loopMenuItem;
exports.loopMenuItemRecusively = loopMenuItemRecusively;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function noop() {}

function getKeyFromChildrenIndex(child, menuEventKey, index) {
  var prefix = menuEventKey || '';
  return child.key || prefix + 'item_' + index;
}

function loopMenuItem(children, cb) {
  var index = -1;
  _react2["default"].Children.forEach(children, function (c) {
    index++;
    if (c && c.type && c.type.isMenuItemGroup) {
      _react2["default"].Children.forEach(c.props.children, function (c2) {
        index++;
        cb(c2, index);
      });
    } else {
      cb(c, index);
    }
  });
}

function loopMenuItemRecusively(children, keys, ret) {
  if (!children || ret.find) {
    return;
  }
  _react2["default"].Children.forEach(children, function (c) {
    if (ret.find) {
      return;
    }
    if (c) {
      var construt = c.type;
      if (!construt || !(construt.isSubMenu || construt.isMenuItem || construt.isMenuItemGroup)) {
        return;
      }
      if (keys.indexOf(c.key) !== -1) {
        ret.find = true;
      } else if (c.props.children) {
        loopMenuItemRecusively(c.props.children, keys, ret);
      }
    }
  });
}