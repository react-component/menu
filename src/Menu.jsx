// import React from 'react';
import PropTypes from 'prop-types';
import createReactClass from 'create-react-class';
import MenuMixin from './MenuMixin';
import { noop } from './util';

const Menu = createReactClass({
  displayName: 'Menu',

  propTypes: {
    defaultSelectedKeys: PropTypes.arrayOf(PropTypes.string),
    selectedKeys: PropTypes.arrayOf(PropTypes.string),
    defaultOpenKeys: PropTypes.arrayOf(PropTypes.string),
    openKeys: PropTypes.arrayOf(PropTypes.string),
    mode: PropTypes.oneOf(['horizontal', 'vertical', 'vertical-left', 'vertical-right', 'inline']),
    getPopupContainer: PropTypes.func,
    onClick: PropTypes.func,
    onSelect: PropTypes.func,
    onDeselect: PropTypes.func,
    onDestroy: PropTypes.func,
    openTransitionName: PropTypes.string,
    openAnimation: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
    subMenuOpenDelay: PropTypes.number,
    subMenuCloseDelay: PropTypes.number,
    forceSubMenuRender: PropTypes.bool,
    triggerSubMenuAction: PropTypes.string,
    level: PropTypes.number,
    selectable: PropTypes.bool,
    multiple: PropTypes.bool,
    children: PropTypes.any,
  },

  mixins: [MenuMixin],

  isRootMenu: true,

  getDefaultProps() {
    return {
      selectable: true,
      onClick: noop,
      onSelect: noop,
      onOpenChange: noop,
      onDeselect: noop,
      defaultSelectedKeys: [],
      defaultOpenKeys: [],
      subMenuOpenDelay: 0,
      subMenuCloseDelay: 0.1,
      triggerSubMenuAction: 'hover',
    };
  },

  getInitialState() {
    const props = this.props;
    let selectedKeys = props.defaultSelectedKeys;
    let openKeys = props.defaultOpenKeys;
    if ('selectedKeys' in props) {
      selectedKeys = props.selectedKeys || [];
    }
    if ('openKeys' in props) {
      openKeys = props.openKeys || [];
    }
    return {
      selectedKeys,
      openKeys,
    };
  },

  componentWillReceiveProps(nextProps) {
    const props = {};
    if ('selectedKeys' in nextProps) {
      props.selectedKeys = nextProps.selectedKeys || [];
    }
    if ('openKeys' in nextProps) {
      props.openKeys = nextProps.openKeys || [];
    }
    this.setState(props);
  },

  onDestroy(key) {
    const state = this.state;
    const props = this.props;
    const selectedKeys = state.selectedKeys;
    const openKeys = state.openKeys;
    let index = selectedKeys.indexOf(key);
    if (!('selectedKeys' in props) && index !== -1) {
      selectedKeys.splice(index, 1);
    }
    index = openKeys.indexOf(key);
    if (!('openKeys' in props) && index !== -1) {
      openKeys.splice(index, 1);
    }
  },

  onSelect(selectInfo) {
    const props = this.props;
    if (props.selectable) {
      // root menu
      let selectedKeys = this.state.selectedKeys;
      const selectedKey = selectInfo.key;
      if (props.multiple) {
        selectedKeys = selectedKeys.concat([selectedKey]);
      } else {
        selectedKeys = [selectedKey];
      }
      if (!('selectedKeys' in props)) {
        this.setState({
          selectedKeys,
        });
      }
      props.onSelect({
        ...selectInfo,
        selectedKeys,
      });
    }
  },

  onClick(e) {
    this.props.onClick(e);
  },

  onOpenChange(e_) {
    const props = this.props;
    const openKeys = this.state.openKeys.concat();
    let changed = false;
    const processSingle = (e) => {
      let oneChanged = false;
      if (e.open) {
        oneChanged = openKeys.indexOf(e.key) === -1;
        if (oneChanged) {
          openKeys.push(e.key);
        }
      } else {
        const index = openKeys.indexOf(e.key);
        oneChanged = index !== -1;
        if (oneChanged) {
          openKeys.splice(index, 1);
        }
      }
      changed = changed || oneChanged;
    };
    if (Array.isArray(e_)) {
      // batch change call
      e_.forEach(processSingle);
    } else {
      processSingle(e_);
    }
    if (changed) {
      if (!('openKeys' in this.props)) {
        this.setState({ openKeys });
      }
      props.onOpenChange(openKeys);
    }
  },

  onDeselect(selectInfo) {
    const props = this.props;
    if (props.selectable) {
      const selectedKeys = this.state.selectedKeys.concat();
      const selectedKey = selectInfo.key;
      const index = selectedKeys.indexOf(selectedKey);
      if (index !== -1) {
        selectedKeys.splice(index, 1);
      }
      if (!('selectedKeys' in props)) {
        this.setState({
          selectedKeys,
        });
      }
      props.onDeselect({
        ...selectInfo,
        selectedKeys,
      });
    }
  },

  getOpenTransitionName() {
    const props = this.props;
    let transitionName = props.openTransitionName;
    const animationName = props.openAnimation;
    if (!transitionName && typeof animationName === 'string') {
      transitionName = `${props.prefixCls}-open-${animationName}`;
    }
    return transitionName;
  },

  isInlineMode() {
    return this.props.mode === 'inline';
  },

  lastOpenSubMenu() {
    let lastOpen = [];
    const { openKeys } = this.state;
    if (openKeys.length) {
      lastOpen = this.getFlatInstanceArray().filter((c) => {
        return c && openKeys.indexOf(c.props.eventKey) !== -1;
      });
    }
    return lastOpen[0];
  },

  renderMenuItem(c, i, subIndex) {
    if (!c) {
      return null;
    }
    const state = this.state;
    const extraProps = {
      openKeys: state.openKeys,
      selectedKeys: state.selectedKeys,
      triggerSubMenuAction: this.props.triggerSubMenuAction,
    };
    return this.renderCommonMenuItem(c, i, subIndex, extraProps);
  },

  render() {
    const props = { ...this.props };
    props.className += ` ${props.prefixCls}-root`;
    return this.renderRoot(props);
  },
});

export default Menu;
