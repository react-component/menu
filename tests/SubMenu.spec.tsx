/* eslint-disable no-undef */
import { act, fireEvent, render } from '@testing-library/react';
import { resetWarned } from 'rc-util/lib/warning';
import React from 'react';
import Menu, { MenuItem, SubMenu } from '../src';
import { isActive, last } from './util';

jest.mock('@rc-component/trigger', () => {
  const react  = require('react');
  let Trigger = jest.requireActual('rc-trigger/lib/mock');
  Trigger = Trigger.default || Trigger;

  return react.forwardRef((props, ref) => {
    global.triggerProps = props;
    return react.createElement(Trigger, { ref, ...props });
  });
});

jest.mock('../src/SubMenu/PopupTrigger', () => {
  const react  = require('react');
  let PopupTrigger = jest.requireActual('../src/SubMenu/PopupTrigger');
  PopupTrigger = PopupTrigger.default || PopupTrigger;

  return react.forwardRef((props, ref) => {
    global.popupTriggerProps = props;
    return react.createElement(PopupTrigger, { ref, ...props });
  });
});

describe('SubMenu', () => {
  function runAllTimer() {
    for (let i = 0; i < 10; i += 1) {
      act(() => {
        jest.runAllTimers();
      });
    }
  }

  beforeEach(() => {
    global.triggerProps = null;
    global.popupTriggerProps = null;
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  function createMenu(props?) {
    return (
      <Menu {...props}>
        <SubMenu key="s1" title="submenu1">
          <MenuItem key="s1-1">1</MenuItem>
          <SubMenu key="s1-2" title="submenu1-1">
            <MenuItem key="s1-2-1">2</MenuItem>
          </SubMenu>
        </SubMenu>
        <SubMenu key="s2" title="submenu2">
          <MenuItem key="s2-2">2</MenuItem>
        </SubMenu>
      </Menu>
    );
  }

  function itemIcon() {
    return <span>MenuItemIcon</span>;
  }

  it("don't show submenu when disabled", () => {
    const { container } = render(
      <Menu mode="vertical">
        <SubMenu key="s" title="submenu" disabled>
          <MenuItem key="1">1</MenuItem>
        </SubMenu>
      </Menu>,
    );
    fireEvent.mouseEnter(container.querySelector('.rc-menu-submenu-title'));

    expect(global.popupTriggerProps.visible).toBeFalsy();
  });

  it('offsets the submenu popover', () => {
    render(
      <Menu mode="horizontal" disabledOverflow>
        <SubMenu key="s" title="submenu" popupOffset={[0, 15]}>
          <MenuItem key="1">1</MenuItem>
        </SubMenu>
      </Menu>,
    );

    const { popupAlign } = global.triggerProps;
    expect(popupAlign).toEqual({ offset: [0, 15] });
  });

  it('should render custom arrow icon correctly.', () => {
    const { container } = render(
      <Menu
        mode="vertical"
        itemIcon={itemIcon}
        expandIcon={<span>SubMenuIconNode</span>}
      >
        <SubMenu key="s" title="submenu">
          <MenuItem key="1">1</MenuItem>
          <MenuItem key="2">2</MenuItem>
        </SubMenu>
      </Menu>,
    );

    const wrapperWithExpandIconFunction = render(
      <Menu
        mode="vertical"
        itemIcon={itemIcon}
        expandIcon={() => <span>SubMenuIconNode</span>}
      >
        <SubMenu key="s" title="submenu">
          <MenuItem key="1">1</MenuItem>
          <MenuItem key="2">2</MenuItem>
        </SubMenu>
      </Menu>,
    );

    const subMenuText = container.querySelector(
      '.rc-menu-submenu-title',
    ).textContent;
    const subMenuTextWithExpandIconFunction =
      wrapperWithExpandIconFunction.container.querySelector(
        '.rc-menu-submenu-title',
      ).textContent;
    expect(subMenuText).toEqual('submenuSubMenuIconNode');
    expect(subMenuTextWithExpandIconFunction).toEqual('submenuSubMenuIconNode');
  });

  it('should Not render custom arrow icon in horizontal mode.', () => {
    const { container } = render(
      <Menu mode="horizontal" disabledOverflow>
        <SubMenu
          key="s"
          title="submenu"
          itemIcon={itemIcon}
          expandIcon={<span>SubMenuIconNode</span>}
        >
          <MenuItem key="1">1</MenuItem>
        </SubMenu>
      </Menu>,
    );

    const childText = container.querySelector(
      '.rc-menu-submenu-title',
    ).textContent;
    expect(childText).toEqual('submenu');
  });

  describe('openSubMenuOnMouseEnter and closeSubMenuOnMouseLeave are true', () => {
    it('toggles when mouse enter and leave', () => {
      const { container } = render(
        <Menu>
          <SubMenu key="s1" title="submenu1">
            <MenuItem key="s1-1">1</MenuItem>
          </SubMenu>
        </Menu>,
      );

      // Enter
      fireEvent.mouseEnter(container.querySelector('.rc-menu-submenu-title'));
      runAllTimer();
      expect(container.querySelector('.rc-menu-submenu-open')).toBeTruthy();

      // Leave
      fireEvent.mouseLeave(container.querySelector('.rc-menu-submenu-title'));
      act(() => {
        jest.runAllTimers();
      });
      expect(container.querySelector('.rc-menu-submenu-open')).toBeFalsy();
    });
  });

  describe('openSubMenuOnMouseEnter and closeSubMenuOnMouseLeave are false', () => {
    it('toggles when mouse click', () => {
      const { container } = render(
        createMenu({
          triggerSubMenuAction: 'click',
        }),
      );

      fireEvent.click(container.querySelector('.rc-menu-submenu-title'));
      runAllTimer();
      expect(container.querySelector('.rc-menu-submenu-open')).toBeTruthy();

      fireEvent.click(container.querySelector('.rc-menu-submenu-title'));
      act(() => {
        jest.runAllTimers();
      });
      expect(container.querySelector('.rc-menu-submenu-open')).toBeFalsy();
    });
  });

  it('fires openChange event', () => {
    const handleOpenChange = jest.fn();
    const { container } = render(
      <Menu onOpenChange={handleOpenChange}>
        <MenuItem key="1">1</MenuItem>
        <SubMenu title="s1">
          <MenuItem key="2">2</MenuItem>
          <SubMenu title="s2">
            <MenuItem key="3">3</MenuItem>
          </SubMenu>
        </SubMenu>
      </Menu>,
    );

    // First
    fireEvent.mouseEnter(container.querySelector('.rc-menu-submenu-title'));
    runAllTimer();
    expect(handleOpenChange).toHaveBeenCalledWith(['tmp_key-1']);

    // Second
    fireEvent.mouseEnter(
      container.querySelectorAll('.rc-menu-submenu-title')[1],
    );
    runAllTimer();
    expect(handleOpenChange).toHaveBeenCalledWith([
      'tmp_key-1',
      'tmp_key-tmp_key-1-1',
    ]);
  });

  describe('undefined key', () => {
    it('warning item', () => {
      resetWarned();

      const errorSpy = jest
        .spyOn(console, 'error')
        .mockImplementation(() => {});

      render(
        <Menu>
          <MenuItem>1</MenuItem>
        </Menu>,
      );

      expect(errorSpy).toHaveBeenCalledWith(
        'Warning: MenuItem should not leave undefined `key`.',
      );

      errorSpy.mockRestore();
    });

    it('warning sub menu', () => {
      resetWarned();

      const errorSpy = jest
        .spyOn(console, 'error')
        .mockImplementation(() => {});

      render(
        <Menu>
          <SubMenu />
        </Menu>,
      );

      expect(errorSpy).toHaveBeenCalledWith(
        'Warning: SubMenu should not leave undefined `key`.',
      );

      errorSpy.mockRestore();
    });

    it('should not warning', () => {
      resetWarned();

      const errorSpy = jest
        .spyOn(console, 'error')
        .mockImplementation(() => {});

      render(
        <Menu>
          <Menu.Divider />
        </Menu>,
      );

      expect(errorSpy).not.toHaveBeenCalled();

      errorSpy.mockRestore();
    });
  });

  describe('mouse events', () => {
    it('mouse enter event on a submenu should not activate first item', () => {
      const { container } = render(createMenu({ openKeys: ['s1'] }));
      fireEvent.mouseEnter(container.querySelector('.rc-menu-submenu-title'));

      runAllTimer();

      expect(container.querySelector('.rc-menu-submenu-open')).toBeTruthy();
      isActive(container, 0, false);
    });

    it('click to open a submenu should not activate first item', () => {
      const { container } = render(
        createMenu({ triggerSubMenuAction: 'click' }),
      );
      fireEvent.click(container.querySelector('.rc-menu-submenu-title'));

      runAllTimer();

      expect(container.querySelector('.rc-menu-submenu-open')).toBeTruthy();
      isActive(container, 0, false);
    });

    it('mouse enter/mouse leave on a subMenu item should trigger hooks', () => {
      const onMouseEnter = jest.fn();
      const onMouseLeave = jest.fn();
      const { container } = render(
        <Menu openKeys={['s1']}>
          <SubMenu
            key="s1"
            title="submenu1"
            onMouseEnter={onMouseEnter}
            onMouseLeave={onMouseLeave}
          >
            <MenuItem key="s1-1">1</MenuItem>
          </SubMenu>
        </Menu>,
      );
      fireEvent.mouseEnter(container.querySelector('.rc-menu-submenu-title'));
      expect(onMouseEnter).toHaveBeenCalledTimes(1);

      fireEvent.mouseLeave(container.querySelector('.rc-menu-submenu-title'));
      expect(onMouseLeave).toHaveBeenCalledTimes(1);
    });
  });

  it('fires select event', () => {
    const handleSelect = jest.fn();
    const { container } = render(createMenu({ onSelect: handleSelect }));
    fireEvent.mouseEnter(container.querySelector('.rc-menu-submenu-title'));

    runAllTimer();

    fireEvent.click(container.querySelector('.rc-menu-item'));
    expect(handleSelect.mock.calls[0][0].key).toBe('s1-1');
  });

  it('fires select event with className', () => {
    const { container } = render(createMenu());
    fireEvent.mouseEnter(container.querySelector('.rc-menu-submenu-title'));

    runAllTimer();

    fireEvent.click(container.querySelector('.rc-menu-item'));
    expect(container.querySelector('.rc-menu-submenu')).toHaveClass(
      'rc-menu-submenu-selected',
    );
  });

  it('fires deselect event for multiple menu', () => {
    const handleDeselect = jest.fn();
    const { container } = render(
      createMenu({
        multiple: true,
        onDeselect: handleDeselect,
      }),
    );
    fireEvent.mouseEnter(container.querySelector('.rc-menu-submenu-title'));

    runAllTimer();

    fireEvent.click(container.querySelector('.rc-menu-item'));
    fireEvent.click(container.querySelector('.rc-menu-item'));
    expect(handleDeselect.mock.calls[0][0].key).toBe('s1-1');
  });

  it('should take style prop', () => {
    const App = () => (
      <Menu style={{ backgroundColor: 'black' }}>
        <SubMenu key="s1" title="submenu1">
          <MenuItem key="s1-1">1</MenuItem>
        </SubMenu>
      </Menu>
    );

    const { container } = render(<App />);
    expect(container.querySelector('.rc-menu')).toHaveStyle({
      backgroundColor: 'black',
    });
  });

  it('not pass style into sub menu item', () => {
    const { container } = render(
      <Menu mode="horizontal" style={{ background: 'green' }} disabledOverflow>
        <MenuItem style={{ color: 'red' }} key="1">
          1
        </MenuItem>
      </Menu>,
    );

    expect(container.querySelector('.rc-menu-item')).toHaveStyle({
      color: 'red',
    });
  });

  it('inline click for open', () => {
    const onOpenChange = jest.fn();

    const { container } = render(
      <Menu mode="inline" onOpenChange={onOpenChange}>
        <SubMenu key="bamboo" title="Bamboo" disabled>
          <MenuItem key="little">Little</MenuItem>
        </SubMenu>
        <SubMenu key="light" title="Light">
          <MenuItem key="sub">Sub</MenuItem>
        </SubMenu>
      </Menu>,
    );

    // Disabled
    fireEvent.click(container.querySelector('.rc-menu-submenu-title'));
    expect(onOpenChange).not.toHaveBeenCalled();

    // Disabled
    fireEvent.click(last(container.querySelectorAll('.rc-menu-submenu-title')));
    expect(onOpenChange).toHaveBeenCalledWith(['light']);
  });

  it('popup className should correct', () => {
    const { container } = render(
      <Menu mode="horizontal" openKeys={['light']} disabledOverflow>
        <SubMenu key="light">
          <SubMenu key="bamboo" />
        </SubMenu>
      </Menu>,
    );

    runAllTimer();

    expect(container.querySelector('.rc-menu-submenu')).toHaveClass(
      'rc-menu-submenu-horizontal',
    );

    expect(last(container.querySelectorAll('.rc-menu-submenu'))).toHaveClass(
      'rc-menu-submenu-vertical',
    );
  });

  it('should support rootClassName', () => {
    const { container } = render(
      <Menu rootClassName="custom-className" defaultOpenKeys={['1', '1-1']}>
        <SubMenu key="1" title="submenu1">
          <MenuItem key="1-1" role="option">
            submenu7
          </MenuItem>
        </SubMenu>
        <MenuItem key="2" role="option">
          2
        </MenuItem>
        <MenuItem key="3" role="option">
          3
        </MenuItem>
      </Menu>,
    );
    expect(container.children).toMatchSnapshot();

    expect(container.querySelector('.rc-menu-root')).toHaveClass(
      'custom-className',
    );
    expect(container.querySelectorAll('.rc-menu-submenu-popup')).toHaveLength(
      0,
    );

    runAllTimer();

    // Enter
    fireEvent.mouseEnter(container.querySelector('.rc-menu-submenu-title'));

    runAllTimer();

    expect(container.querySelector('.rc-menu-submenu-popup')).toHaveClass(
      'custom-className',
    );

    expect(container.children).toMatchSnapshot();
  });
});
/* eslint-enable */
