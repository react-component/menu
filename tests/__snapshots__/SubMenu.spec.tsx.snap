// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubMenu should support rootClassName 1`] = `
HTMLCollection [
  <ul
    class="rc-menu rc-menu-root rc-menu-vertical custom-className"
    data-menu-list="true"
    role="menu"
    tabindex="0"
  >
    <li
      class="rc-menu-submenu rc-menu-submenu-vertical rc-menu-submenu-open"
      role="none"
    >
      <div
        aria-controls="rc-menu-uuid-test-1-popup"
        aria-expanded="true"
        aria-haspopup="true"
        class="rc-menu-submenu-title"
        data-menu-id="rc-menu-uuid-test-1"
        role="menuitem"
        tabindex="-1"
        title="submenu1"
      >
        submenu1
        <i
          class="rc-menu-submenu-arrow"
        />
      </div>
    </li>
    <li
      aria-selected="false"
      class="rc-menu-item"
      data-menu-id="rc-menu-uuid-test-2"
      role="option"
      tabindex="-1"
    >
      2
    </li>
    <li
      aria-selected="false"
      class="rc-menu-item"
      data-menu-id="rc-menu-uuid-test-3"
      role="option"
      tabindex="-1"
    >
      3
    </li>
  </ul>,
  <div
    aria-hidden="true"
    style="display: none;"
  />,
]
`;

exports[`SubMenu should support rootClassName 2`] = `
HTMLCollection [
  <ul
    class="rc-menu rc-menu-root rc-menu-vertical custom-className"
    data-menu-list="true"
    role="menu"
    tabindex="0"
  >
    <li
      class="rc-menu-submenu rc-menu-submenu-vertical rc-menu-submenu-open rc-menu-submenu-active"
      role="none"
    >
      <div
        aria-controls="rc-menu-uuid-test-1-popup"
        aria-expanded="true"
        aria-haspopup="true"
        class="rc-menu-submenu-title"
        data-menu-id="rc-menu-uuid-test-1"
        role="menuitem"
        tabindex="-1"
        title="submenu1"
      >
        submenu1
        <i
          class="rc-menu-submenu-arrow"
        />
      </div>
      <div
        class="rc-menu-submenu rc-menu-submenu-popup custom-className rc-menu-submenu-placement-rightTop"
        style="--arrow-x: 0px; --arrow-y: 0px; left: -1000vw; top: -1000vh; box-sizing: border-box;"
      >
        <ul
          class="rc-menu rc-menu-sub rc-menu-vertical"
          data-menu-list="true"
          id="rc-menu-uuid-test-1-popup"
          role="menu"
        >
          <li
            aria-selected="false"
            class="rc-menu-item"
            data-menu-id="rc-menu-uuid-test-1-1"
            role="option"
            tabindex="-1"
          >
            submenu7
          </li>
        </ul>
      </div>
    </li>
    <li
      aria-selected="false"
      class="rc-menu-item"
      data-menu-id="rc-menu-uuid-test-2"
      role="option"
      tabindex="-1"
    >
      2
    </li>
    <li
      aria-selected="false"
      class="rc-menu-item"
      data-menu-id="rc-menu-uuid-test-3"
      role="option"
      tabindex="-1"
    >
      3
    </li>
  </ul>,
  <div
    aria-hidden="true"
    style="display: none;"
  />,
]
`;
